cli/loader/src/lib.rs

```diff
@@ -532,45 +533,94 @@ impl Loader {
         output_path: &PathBuf,
         force_docker: bool,
     ) -> Result<(), Error> {
-        let emcc_bin = if cfg!(windows) { "emcc.bat" } else { "emcc" };
-        let emcc_path = which(emcc_bin)
-            .ok()
-            .and_then(|p| Command::new(&p).output().and(Ok(p)).ok());
-
-        let mut command;
-        if emcc_path.is_some() && !force_docker {
-            command = Command::new(emcc_path.unwrap());
-            command.current_dir(&src_path);
-        } else if Command::new("docker").output().is_ok() {
-            command = Command::new("docker");
-            command.args(&["run", "--rm"]);
-
-            // Mount the parser directory as a volume
-            command.args(&["--workdir", "/src"]);
-
-            let mut volume_string = OsString::from(&src_path);
-            volume_string.push(":/src:Z");
-            command.args(&[OsStr::new("--volume"), &volume_string]);
-
-            // Get the current user id so that files created in the docker container will have
-            // the same owner.
-            if cfg!(unix) {
-                let user_id_output = Command::new("id")
-                    .arg("-u")
-                    .output()
-                    .with_context(|| "Failed to get get current user id")?;
-                let user_id = String::from_utf8_lossy(&user_id_output.stdout);
-                let user_id = user_id.trim();
-                command.args(&["--user", user_id]);
-            }
-
-            // Run `emcc` in a container using the `emscripten-slim` image
-            command.args(&[EMSCRIPTEN_TAG, "emcc"]);
-        } else {
+        enum EmscriptenSource {
+            Native(PathBuf),
+            Docker,
+            Podman,
+        }
+        // prefer emscripten > docker > podman > error
+        let exe: Option<EmscriptenSource> = {
+            let emcc: Option<_> = if !force_docker {
+                let emcc_bin = if cfg!(windows) { "emcc.bat" } else { "emcc" };
+                let emcc_path = which(emcc_bin)
+                    .ok()
+                    .and_then(|p| Command::new(&p).output().and(Ok(p)).ok());
+                emcc_path.map(|p| EmscriptenSource::Native(p))
+            } else {
+                None
+            };
+            emcc.or_else(|| {
+                if Command::new("docker").output().is_ok() {
```

Comment:
I'm wondering if we should fall back to `podman` even if `docker` is installed, if the Docker server isn't running. To do that, we could run `docker images` instead of just `docker`, so that we'll get a failure if docker isn't running.

Reply:
I pushed this behavior in 61f7c2bc. For a future improvement, we might surface information about what executables are present on `$PATH` or what errors each executable raises.